netutils/dhcp6c/dhcp6c.c

```diff
@@ -0,0 +1,1951 @@
+    {
+      if (otype == DHCPV6_OPT_CLIENTID)
+        {
+          clientid_ok = (olen + 4U == client_id_len) &&
+                        !memcmp((odata - 4), client_id, client_id_len);
+        }
+      else if (otype == DHCPV6_OPT_SERVERID)
+        {
+          serverid_ok = (olen + 4U == server_id_len) &&
+                        !memcmp((odata - 4), server_id, server_id_len);
+        }
+    }
+
+  return clientid_ok && (serverid_ok || server_id_len == 0);
+}
+
+static int dhcp6c_command(FAR void *handle, enum dhcpv6_msg_e type)
+{
+  FAR struct dhcp6c_state_s *pdhcp6c = (FAR struct dhcp6c_state_s *)handle;
+  uint8_t buf[1536];
+  uint32_t timeout = CONFIG_NETUTILS_DHCP6C_REQUEST_TIMEOUT < 3? 3 :
+                     CONFIG_NETUTILS_DHCP6C_REQUEST_TIMEOUT;
+  FAR const struct dhcp6c_retx_s *retx = &g_dhcp6c_retx[type];
+  uint64_t start;
+  uint64_t round_start;
+  uint64_t round_end;
+  uint64_t elapsed;
+  uint8_t trid[3];
+  ssize_t len = -1;
+  int64_t rto = 0;
+
+  if (retx->delay)
+    {
+      struct timespec ts;
+      ts.tv_sec = 0;
+      ts.tv_nsec = dhcp6c_rand_delay(handle, 10 * DHCPV6_REQ_DELAY);
+      nanosleep(&ts, NULL);
+    }
+
+  if (type == DHCPV6_MSG_RELEASE || type == DHCPV6_MSG_DECLINE)
+    {
+      timeout = 3;
+    }
+  else if (type == DHCPV6_MSG_UNKNOWN)
+    {
+      timeout = pdhcp6c->t1;
+    }
+  else if (type == DHCPV6_MSG_RENEW)
+    {
+      timeout = pdhcp6c->t2 - pdhcp6c->t1;
+    }
+  else if (type == DHCPV6_MSG_REBIND)
+    {
+      timeout = pdhcp6c->t3 - pdhcp6c->t2;
+    }
+
+  if (timeout == 0)
+    {
+      return -1;
+    }
+
+  ninfo("Sending %s (timeout %u s)\n", retx->name, timeout);
+  start = dhcp6c_get_milli_time();
+  round_start = start;
+
+  /* Generate transaction ID */
+
+  read(pdhcp6c->urandom_fd, trid, sizeof(trid));
+
+  do
+    {
+      rto = (rto == 0) ? (retx->init_timeo * 1000 +
+              dhcp6c_rand_delay(handle, retx->init_timeo * 1000)) :
+              (2 * rto + dhcp6c_rand_delay(handle, rto));
+
+      if (rto >= retx->max_timeo * 1000)
+        {
+          rto = retx->max_timeo * 1000 +
+                dhcp6c_rand_delay(handle, retx->max_timeo * 1000);
+        }
+
+      /* Calculate end for this round and elapsed time */
+
+      round_end = round_start + rto;
+      elapsed = round_start - start;
+
+      /* Don't wait too long */
+
+      if (round_end - start > timeout * 1000)
+        {
+          round_end = timeout * 1000 + start;
+        }
+
+      /* Built and send package */
+
+      if (type != DHCPV6_MSG_UNKNOWN)
+        {
+          dhcp6c_send(handle, type, trid, elapsed / 10);
+        }
+
+      /* Receive rounds */
+
+      for (; len < 0 && round_start < round_end;
+              round_start = dhcp6c_get_milli_time())
+        {
+          /* Set timeout for receiving */
+
+          uint64_t t = round_end - round_start;
+          struct timeval retime =
+          {
+            t / 1000, (t % 1000) * 1000
```

Comment:
```suggestion
            t / MSEC_PER_SEC, (t % MSEC_PER_SEC) * USEC_PER_MSEC
```
here and in other places

Reply:
@pkarashchenko since xSEC_PER_SEC is NuttX specific macro, it's better to avoid use them to improve dhcpv6 portability.

Reply:
We can do it one by one. For general application, I don't see any reason why not make them could run on POSIX compliant OS(e.g. Linux) too.